src/webgpu/api/operation/command_buffer/render/state_tracking.spec.ts

```diff
@@ -18,3 +17,163 @@ import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+      vertex: {
+        module: t.device.createShaderModule({
+          code: `
+            struct Inputs {
+              [[location(0)]] vertexPosition : f32;
+              [[location(1)]] vertexColor : vec4<f32>;
+            };
+            struct Outputs {
+              [[builtin(position)]] position : vec4<f32>;
+              [[location(0)]] color : vec4<f32>;
+            };
+            [[stage(vertex)]]
+            fn main(input : Inputs)-> Outputs {
+              var outputs : Outputs;
+              outputs.position =
+                vec4<f32>(input.vertexPosition, 0.5, 0.0, 1.0);
+              outputs.color = input.vertexColor;
+              return outputs;
+            }`,
+        }),
+        entryPoint: 'main',
+        buffers: [
+          {
+            arrayStride: kVertexAttributeSize,
+            attributes: [
+              {
+                format: 'float32',
+                offset: 0,
+                shaderLocation: 0,
+              },
+              {
+                format: 'unorm8x4',
+                offset: 4,
+                shaderLocation: 1,
+              },
+            ],
+          },
+        ],
+      },
+      fragment: {
+        module: t.device.createShaderModule({
+          code: `
+            struct Input {
+              [[location(0)]] color : vec4<f32>;
+            };
+            [[stage(fragment)]]
+            fn main(input : Input) -> [[location(0)]] vec4<f32> {
+              return input.color;
+            }`,
+        }),
+        entryPoint: 'main',
+        targets: [{ format: 'rgba8unorm' }],
+      },
+      primitive: {
+        topology: 'point-list',
+      },
+    });
+
+    const outputTexture = t.device.createTexture({
+      format: 'rgba8unorm',
+      size: [kPositions.length - 1, 1, 1],
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: outputTexture.createView(),
+          loadValue: [0, 0, 0, 1],
+          storeOp: 'store',
+        },
+      ],
+    });
+    renderPass.setPipeline(renderPipeline);
+    renderPass.setVertexBuffer(0, vertexBuffer);
+
+    // 1st draw: indexFormat = 'uint32', offset = 0, size = 4 (index value: 0x10000)
+    renderPass.setIndexBuffer(indexBuffer, 'uint32', 0, 4);
+    renderPass.drawIndexed(1);
+
+    // 2nd draw: indexFormat = 'uint16', offset = 0, size = 2 (index value: 0)
+    renderPass.setIndexBuffer(indexBuffer, 'uint16', 0, 2);
+    renderPass.drawIndexed(1);
+
+    // 3rd draw: indexFormat = 'uint16', offset = 4, size = 2 (index value: 2)
+    renderPass.setIndexBuffer(indexBuffer, 'uint16', 4, 2);
+    renderPass.drawIndexed(1);
+
+    // 4th draw: indexformat = 'uint16', offset = 6, size = 4 (index values: 3, 4)
+    renderPass.setIndexBuffer(indexBuffer, 'uint16', 6, 4);
+    renderPass.drawIndexed(2);
```

Comment:
Since the goal is to make sure that any minimal change in the index state takes effect, how about these small tweaks:

```suggestion
    // 1st draw: indexFormat = 'uint32', offset = 0, size = 4 (index value: 0x10000)
    renderPass.setIndexBuffer(indexBuffer, 'uint32', 0, 4);
    renderPass.drawIndexed(1);

    // 2nd draw: indexFormat = 'uint16', offset = 0, size = 2 (index value: 0)
    renderPass.setIndexBuffer(indexBuffer, 'uint16', 0, 4);
    renderPass.drawIndexed(1);

    // 3rd draw: indexFormat = 'uint16', offset = 4, size = 2 (index value: 2)
    renderPass.setIndexBuffer(indexBuffer, 'uint16', 0, 2);
    renderPass.setIndexBuffer(indexBuffer, 'uint16', 4, 2);
    renderPass.drawIndexed(1);

    // 4th draw: indexformat = 'uint16', offset = 6, size = 4 (index values: 3, 4)
    renderPass.setIndexBuffer(indexBuffer, 'uint16', 6, 2);
    renderPass.setIndexBuffer(indexBuffer, 'uint16', 6, 4);
    renderPass.drawIndexed(2);
```